README
------------------------------------------------------------------------------
                              libburnia-project.org
------------------------------------------------------------------------------
This all is under GPL.
(See GPL reference, our clarification and commitment at the end of this text)
------------------------------------------------------------------------------
libburnia-project.org
By Mario Danic <mario.danic@gmail.com> and Thomas Schmitt <scdbackup@gmx.net>
Copyright (C) 2006-2021 Mario Danic, Thomas Schmitt
Still containing parts of Libburn. By Derek Foreman <derek@signalmarketing.com>
and Ben Jansens <xor@orodu.net>
Copyright (C) 2002-2006  Derek Foreman and Ben Jansens

     http://files.libburnia-project.org/releases/libburn-1.5.4.tar.gz

------------------------------------------------------------------------------

                           Build and Installation

                                From tarball

Obtain libburn-1.5.4.tar.gz, take it to a directory of your choice and do:

   tar xzf libburn-1.5.4.tar.gz
   cd libburn-1.5.4
   ./configure --prefix=/usr
   make

To make libburn accessible for running and application development,
and to install the cdrecord compatibility binary cdrskin, do
(as Superuser):

   make install

This procedure installs libburn.so.4 and cdrskin depending on it.
For a standalone cdrskin binary, see cdrskin/README.

A behavioral conflict is known between any burn software and demons like hald
which probe CD drives. This can spoil burn runs for CD-R or CD-RW.
You may have to keep your hald away from the drive. See for example
   http://www.freebsd.org/gnome/docs/halfaq.html


                                  From SVN

Our build system is based on autotools. For preparing the build of a SVN 
snapshot you will need autotools of at least version 1.7. 
Do in a directory of your choice:

   svn co http://svn.libburnia-project.org/libburn/trunk libburn-svn
   cd libburn-svn
   ./bootstrap
   ./configure --prefix=/usr
   make
   make install

Warning: The trunk might contain experimental features which might not
         persist until next release. 


                        Special ./configure options

make install on GNU/Linux will try to run program ldconfig with the library
installation directory as only argument. Failure to do so will not abort
installation. One may disable ldconfig by ./configure option:
   --disable-ldconfig-at-install

In some situations Linux may deliver a better write performance to drives if
the track input is read with O_DIRECT (see man 2 open). The API call
burn_os_open_track_src() and the input readers of cdrskin and libburn fifo
can be told to use this peculiar read mode by:
   --enable-track-src-odirect

But often libburn call burn_write_opts_set_dvd_obs(opts, 64*1024) will yield
even better performance in such a situation. 64k can be made default at
configure time by:
    --enable-dvd-obs-64k
This may be combined with above --enable-track-src-odirect .

If it is desired that DVD DAO writing and stdio: writing get padded up to
a full write chunk of 32k or 64k, then use ./configure option:
    --enable-dvd-obs-pad

Alternatively the transport of SCSI commands can be done via libcdio-0.83.
You may install it and re-run libburn's ./configure with option
    --enable-libcdio

By use of a version script, the libburn.so library exposes no other function
names but those of the API definition in libburn/libburn.h.
If -Wl,--version-script=... makes problems with the local compiler, then
disable this encapsulation feature by
    --disable-versioned-libs

Make sure to re-compile all source files after running ./configure
   make clean ; make
   make install


Linux only:

libburn tries to avoid a collision with udev's drive examination by waiting
0.1 seconds before opening the device file for a longer time, after udev
might have been alarmed by drive scanning activities.
The waiting time can be set at ./configure time with microsecond granularity.
E.g. 2 seconds:
  CFLAGS="$CFLAGS -DLibburn_udev_wait_useC=2000000"
  ./configure ...options...
Waiting can be disabled by zero waiting time:
  CFLAGS="$CFLAGS -DLibburn_udev_wait_useC=0"
Alternatively, libburn can try to be nice by opening the device file,
closing it immediately, waiting, and only then opening it for real:
  CFLAGS="$CFLAGS -DLibburn_udev_extra_open_cyclE -DLibburn_udev_wait_useC=500000"

------------------------------------------------------------------------------

An important part of the project, libisofs, is hosted in a bzr repository at
launchpad.net :
   bzr branch lp:libisofs

Another part the project, libisoburn, is hosted in the libburnia SVN, too:
   svn co http://svn.libburnia-project.org/libisoburn/trunk libisoburn

See README files there.

------------------------------------------------------------------------------


                        Overview of libburnia-project.org

libburnia-project.org is an open-source software project for reading, mastering
and writing optical discs.
For now this means CD media, all DVD media, all BD media.

The project comprises of several more or less interdependent parts which
together strive to be a usable foundation for application development.
These are libraries, language bindings, and middleware binaries which emulate
classical (and valuable) Linux tools.
Currently it is supported on GNU/Linux with kernels >= 2.4,
on FreeBSD with ATAPI/CAM enabled in the kernel (see man atapicam),
on OpenSolaris (tested with kernel 5.11),
on NetBSD (tested with 6.1.3).
On other X/Open compliant systems there will only be pseudo drives, but no
direct MMC operation on real CD/DVD/BD drives.

For full ports to other systems we would need : login on a development machine
or a live OS on CD or DVD, advise from a system person about the equivalent
of Linux sg or FreeBSD CAM, volunteers for testing of realistic use cases.

We have a well tested code base for burning data and audio CDs, DVDs and BDs.
The burn API is quite comprehensively documented and can be used to build a
presentable application.
We have a functional application which emulates the core use cases of cdrecord
in order to prove that usability, and in order to allow you to explore
libburn's scope by help of existing cdrecord frontends.

ISO 9660 filesystems with Rock Ridge and Joliet extensions can be created
and manipulated quite freely. This capability together with our burn capability
makes possible a single binary application which covers all steps of image
composition, updating and writing. Quite unique in the Linux world.

The project components (list subject to growth, hopefully):

- libburn  is the library by which preformatted data get onto optical media.
           It uses either /dev/sgN (e.g. on kernel 2.4 with ide-scsi) or
           /dev/srM or /dev/hdX (e.g. on kernel 2.6).
           libburn is the foundation of our cdrecord emulation. Its code is
           independent of cdrecord. Its DVD capabilities are learned from
           studying the code of dvd+rw-tools and MMC-5 specs. No code but only
           the pure SCSI knowledge has been taken from dvd+rw-tools, though.

- libisofs is the library to pack up hard disk files and directories into a
           ISO 9660 disk image. This may then be brought to CD via libburn.
           An own ISO 9660 extension stores ACLs, xattr, and MD5 of file
           content.

- libisoburn is an add-on to libburn and libisofs which coordinates both and
           also can grow ISO-9660 filesystem images on multi-session media
           as well as on overwriteable media via the same API.
           All media peculiarities are handled automatically.
           It also contains the methods of command oriented application
           xorriso and offers them via a C language API.

- cdrskin  is a limited cdrecord compatibility wrapper for libburn.
           cdrecord is a powerful GPL'ed burn program included in Joerg
           Schilling's cdrtools. cdrskin strives to be a second source for
           the services traditionally provided by cdrecord. Additionally it
           provides libburn's DVD capabilities, where only -sao is compatible
           with cdrecord.
           cdrskin does not contain any bytes copied from cdrecord's sources.
           Many bytes have been copied from the message output of cdrecord
           runs, though.
           See cdrskin/README for more.

- xorriso  is an application of all three libraries which creates, loads,
           manipulates and writes ISO 9660 filesystem images with
           Rock Ridge extensions. Manipulation is not only adding or
           overwriting of files but also deleting, renaming, attribute
           changing, incremental backups, activating boot images, and
           extracting of files from ISO images to disk. There is also a
           sparse emulation of cdrecord and a more laborate one of mkisofs.
           All features of xorriso are also available via a C language API
           of libisoburn.
           A static compilation of xorriso and the libraries is dedicated
           to the GNU Operating System. See xorriso/README_gnu_xorriso .

- "test"   is a collection of application gestures and examples given by the
           authors of the library features. The burn API example of libburn
           is named test/libburner.c . The API for media information inquiry is
           demonstrated in test/telltoc.c .
           Explore these examples if you look for inspiration.

We strive to be a responsive upstream.

Our libraries are committed to maintain older feature sets in newer versions.
This applies to source code headers (API) as well as to linkable objects (ABI).
The only exception from this rule is about non-release versions x.y.*[13579]
which are allowed to introduce new features, change those new features in
any way and even may revoke such new features before the next release of
x.y.*[02468]. As soon as it is released, a feature is promised to persist.

SONAMES:
libburn.so.4    (since 0.3.4, March    2007),
libisofs.so.6   (since 0.6.2, February 2008),
libisoburn.so.1 (since 0.1.0, February 2008).

Applications must use 64 bit off_t. E.g. by defining
#define _LARGEFILE_SOURCE
#define _FILE_OFFSET_BITS 64
or take special precautions to interface with the libraries by 64 bit integers
where the .h files prescribe off_t. To reduce libburn's off_t size to 32 bit
will keep it from processing tracks of more than 2 GB size.


------------------------------------------------------------------------------
Project history as far as known to me:

- Founded in 2002 as it seems. See mailing list archives
     http://lists.freedesktop.org/archives/libburn/
  The site of this founder team is reachable and offers download of a
  (somewhat outdated) tarball and from CVS :
     http://icculus.org/burn/
  Copyright holders and most probably founders:
  Derek Foreman and Ben Jansens.

- I came to using libburn in 2005. Founded the cdrskin project and submitted
  necessary patches which were accepted or implemented better. Except one
  remaining patch which prevented cdrskin from using vanilla libburn from CVS.
  The cdrskin project site is reachable and offers download of the heavily
  patched (elsewise outdated) tarball under the name  cdrskin-0.1.2  :
     http://scdbackup.sourceforge.net/cdrskin_eng.html
  It has meanwhile moved to use vanilla libburn.pykix.org , though.
  Version 0.1.4 constitutes the first release of this kind.

- In July 2006 our team mate Mario Danic announced a revival of libburn
  which by about nearly everybody else was perceived as unfriendly fork.
  Derek Foreman four days later posted a message which expressed his
  discontent.
  The situation first caused me to publicly regret it and then - after i
  got the opportunity to move in with cdrskin - gave me true reason to
  personally apologize to Derek Foreman, Ben Jansens and the contributors at 
  icculus.org/burn. Posted to both projects:
    http://lists.freedesktop.org/archives/libburn/2006-August/000446.html
    http://mailman-mail1.webfaction.com/pipermail/libburn-hackers/2006-August/000024.html

- Mid August 2006 project cdrskin established a branch office in 
  libburn.pykix.org so that all maintainers of our tools have one single place
  to get the current (at least slightely) usable coordinated versions of
  everything.
  Project cdrskin will live forth independendly for a while but it is committed
  to stay in sync with libburn.pykix.org (or some successor, if ever).
  cdrskin is also committed to support  icculus.org/burn  if the pending fork
  is made reality by content changes in that project. It will cease to maintain
  a patched version of  icculus.org/burn  though. Precondition for a new
  release of cdrskin on base of  icculus.org/burn  would be the pending
  "whitelist patch" therefore.
  I would rather prefer if both projects find consense and merge, or at least
  cooperate. I have not given up hope totally, yet.
  I, personally, will honor any approach.

- 2nd September 2006 the decision is made to strive for a consolidation of
  copyright and a commitment to GPL in a reasonable and open minded way.
  This is to avoid long term problems with code of unknown origin and
  with finding consense among the not so clearly defined group of copyright
  claimers and -holders.
  libisofs is already claimed sole copyright Mario Danic.
  cdrskin and libburner are already claimed sole copyright Thomas Schmitt.
  Rewrites of other components will follow and concluded by claiming full
  copyright within the group of libburn.pykix.org-copyright holders.

- 16th September 2006 feature freeze for release of libburn-0.2.2 .

- 20th September 2006 release of libburn-0.2.2 .

- 26th October 2006 feature freeze for cdrskin-0.2.4 based on libburn-0.2.3 .
  This version of cdrskin is much more cdrecord compatible in repect
  to drive addressing and audio features.

- 30th October 2006 release of cdrskin-0.2.4 .

- 13th November 2006 splitting releases of libburn+cdrskin from libisofs.

- 24th November 2006 release of libburn-0.2.6 and cdrskin-0.2.6 . cdrskin has
  become suitable for unaware frontends as long as they perform only the core 
  of cdrecord use cases (including open-ended input streams, audio, and
  multi-session).

- 28th November 2006 the umbrella project which encloses both, libisofs and
  libburn, is now called libburnia. For the origin of this name, see 
  http://en.wikipedia.org/wiki/Liburnians .

- 16th January 2007 release of libburn-0.3.0 and cdrskin-0.3.0 . Now the scope
  is widened to a first class of DVD media: overwriteable single layer types
  DVD-RAM, DVD+RW, DVD-RW. This is not a cdrecord emulation but rather inspired
  by dvd+rw-tools' "poor man" writing facility for this class of media.
  Taking a bow towards Andy Polyakov.

- 11th February 2007 version 0.3.2 covers sequential DVD-RW and DVD-R with
  multi-session and with DAO.

- 12th March 2007 version 0.3.4 supports DVD+R and thus covers all single layer
  DVD media. Code for double layer DVD+/-R is implemented but awaits a tester
  yet.

- 23th April 2007 version 0.3.6 follows the unanimous opinion of Linux kernel
  people that one should not use /dev/sg on kernel 2.6.


ChangeLog
git clone git@dev.lovelyhq.com:libburnia/libburn.git
(to become libburn-1.5.4 or higher)
===============================================================================
- no novelties yet -

libburn-1.5.2.tar.gz Sat Oct 26 2019
===============================================================================
* Bug fix: No lock was obtained for setting up a fifo object
* Bug fix: Stream recording was applied regardless whether the drive offers it.
           This caused Xfburn failures with some MATSHITA laptop drives.
* Bug fix: TDK Corporation was not recognized as manufacturer of DVD-R "TTH02"
* Made libburn ready for building out-of-source.  Thanks Ross Burton.
* New API calls burn_drive_get_feature_codes(), burn_drive_get_feature()
* New cdrskin option --list_features

libburn-1.5.0.tar.gz Sat Sep 15 2018
===============================================================================
* Bug fix: cdrskin threw errno 22 on data file input if libburn is
           configured with --enable-track-src-odirect
* Bug fix: SIGSEGV could happen if a track ended by reaching its fixed size
           while the track source still was willing to deliver bytes.
           Thanks to user swordragon.
* Bug fix: Device file comparison parameters were recorded wrong with Linux sg

libburn-1.4.8.tar.gz Tue Sep 12 2017
===============================================================================
* Bug fix: Option -dummy did not affect writing by direct_write_amount=
* New API call burn_drive_reset_simulate()
* New API call burn_drive_get_bd_r_pow()
* Refusing to write to BD-R if formatted to Pseudo Overwrite

libburn-1.4.6.tar.gz Fri Sep 16 2016
===============================================================================
* Bug fix: SAO CD could be perceived 2 blocks to short.
           Regression in 1.4.4 by rev 5672.
* Now operating optical drives on OpenBSD. Thanks to SASANO Takayoshi.
* New API call burn_drive_set_immed()
* New cdrskin option use_immed_bit=

libburn-1.4.4.tar.gz Fri Jul 01 2016
===============================================================================
* Bug fix: Option drive_scsi_dev_family=sg did not convert /dev/sr* to /dev/sg*
* Bug fix: burn_make_input_sheet_v07t() falsly recognized double byte encoding.
           Affected cdrskin option: cdtext_to_v07t=
* Bug fix: Double free at end of run if burn_write_opts_set_leadin_text() is
           used. Affected cdrskin option: textfile=
* Bug fix: DVD book type of DVD+RW DL and DVD+R DL was reported wrong.
           Thanks to Etienne Bergeron.

libburn-1.4.2.pl01.tar.gz Fri Jan 29 2016
===============================================================================
* Bug fix: cdrskin "failed to attach fifo" when burning from stdin. 
           Regression of 1.4.2, rev 5522.

libburn-1.4.2.tar.gz Sat Nov 28 2015
===============================================================================
* Bug fix: burn_disc_get_media_id() returned BD identifiers 2 chars too long
* Bug fix: burn_disc_get_multi_caps() returned 2048 bytes too many in
           caps.start_range_high
* Bug fix: Media summary session count of blank and closed media was short by 1
* Bug fix: Endless loop if transport error occurs while waiting for drive ready
* New API calls burn_drive_get_serial_no() and burn_drive_get_media_sno()
* Result of a Coverity audit: 40+ code changes, but no easy-to-trigger bugs

libburn-1.4.0.tar.gz Sun May 17 2015
===============================================================================
* Bug fix: Double free with cdrskin -vvv.
           Introduced with rev 5065, version 1.3.1
* Bug fix: Wrong read access to memory. Reported by valgrind of lian jianfei.

libburn-1.3.8.tar.gz Sat Jun 28 2014
===============================================================================
* Bug fix: Wrong stack usage caused SIGBUS on sparc when compiled by gcc -O2
* Bug fix: Minimum drive buffer fill was measured by cdrskin before the buffer
           could get full
* Bug fix: A failed MMC BLANK command did not cause error indication by libburn
* Bug fix: A final fsync(2) was performed with stdio drives, even if not
           desired

libburn-1.3.6.pl01.tar.gz Tue Mar 18 2013
===============================================================================
* Bug fix: CD TAO with multiple tracks could cause a buffer overrun
* Bug fix: Compilation warning for unsupported systems mutated into an error

libburn-1.3.6.tar.gz Tue Mar 04 2013
===============================================================================
* New system adapter for NetBSD

libburn-1.3.4.tar.gz Thu Dec 12 2013
===============================================================================
* Bug fix: Drive error reports were ignored during blanking and formatting
* Bug fix: Drive LG BH16NS40 stalls on inspection of unformatted DVD+RW
* New API call burn_disc_pretend_full_uncond()

libburn-1.3.2.tar.gz Wed Aug 07 2013
===============================================================================
* Bug fix: cdrskin -msinfo on DVD and BD reported
           old session start = next writable address.
           Regression introduced by version 1.2.8 (rev 4956).
* Bug fix: The signal handler aborted on SIGCONT, SIGTSTP, SIGTTIN, SIGTTOU
* New API call burn_make_input_sheet_v07t()
* API call burn_session_input_sheet_v07t(): read multiple blocks from same file
* New API calls burn_drive_extract_audio(), burn_drive_extract_audio_track()
* New cdrskin option textfile_to_v07t=
* New cdrskin options cdtext_to_textfile= and cdtext_to_v07t=
* New cdrskin options extract_audio_to= , extract_tracks= , extract_basename= ,
  --extract_dap
* New cdrskin option --pacifier_with_newline
* Improved granularity of SCSI log time measurement, now with timestamp
* Optional "make doc" now demands doxygen 1.8.4

libburn-1.3.0.pl01.tar.gz Fri May 31 2013
===============================================================================
* Bug fix: cdrskin -msinfo on DVD and BD reported

libburn/util.c

/* Copyright (c) 2004 - 2006 Derek Foreman, Ben Jansens
   Copyright (c) 2006 - 2013 Thomas Schmitt <scdbackup@gmx.net>
   Provided under GPL version 2 or later.
*/

#ifdef HAVE_CONFIG_H
#include "../config.h"
#endif

#include <string.h>

/* ts A61008 */
/* #include <a ssert.h> */

#include <stdlib.h>
#include <stdio.h>
#include <time.h>
#include <sys/time.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>


/* ts A80914 : This is unneeded. Version info comes from libburn.h.
#include "v ersion.h"
*/

#include "util.h"
#include "libburn.h"


void burn_version(int *major, int *minor, int *micro)
{
/* ts A80408 : switched from configure.ac versioning to libburn.h versioning */
	*major = burn_header_version_major;
	*minor = burn_header_version_minor;
	*micro = burn_header_version_micro;
}


struct cd_mid_record {
	char *manufacturer;
	int m_li;
	int s_li;
	int f_li;
	char *other_brands;
};
typedef struct cd_mid_record cd_mid_record_t;


/* ts A90902 */
/** API
   @param flag  Bitfield for control purposes,
                bit0= append "(aka %s)",other_brands to reply
 */
char *burn_guess_cd_manufacturer(int m_li, int s_li, int f_li,
                                 int m_lo, int s_lo, int f_lo, int flag)
{
	static cd_mid_record_t mid_list[]= {
	{"SKC",                                 96, 40,  0,  ""},
	{"Ritek Corp"                         , 96, 43, 30,  ""},
	{"TDK / Ritek"                        , 97, 10,  0,  "TRAXDATA"},
	{"TDK Corporation"                    , 97, 15,  0,  ""},
	{"Ritek Corp"                         , 97, 15, 10,  "7-plus, Aopen, PONY, Power Source, TDK, TRAXDATA, HiCO, PHILIPS, Primdisc, Victor.JVC, OPTI STORAGE, Samsung"},
	{"Mitsubishi Chemical Corporation"    , 97, 15, 20,  ""},
	{"Nan-Ya Plastics Corporation"        , 97, 15, 30,  "Hatron, MMore, Acer, LITEON"},
	{"Delphi"                             , 97, 15, 50,  ""},
	{"Shenzhen SG&SAST"                   , 97, 16, 20,  ""},
	{"Moser Baer India Limited"           , 97, 17,  0,  "EMTEC, Intenso, YAKUMO, PLATINUM, Silver Circle"},
	{"SKY media Manufacturing SA"         , 97, 17, 10,  ""},
	{"Wing"                               , 97, 18, 10,  ""},
	{"DDT"                                , 97, 18, 20,  ""},
	{"Daxon Technology Inc. / Acer"       , 97, 22, 60,  "Maxmax, Diamond Data, BenQ, gold, SONY"},
	{"Taiyo Yuden Company Limited"        , 97, 24,  0,  "Maxell, FUJIFILM, SONY"},
	{"Sony Corporation"                   , 97, 24, 10,  "LeadData, Imation"},
	{"Computer Support Italcard s.r.l"    , 97, 24, 20,  ""},
	{"Unitech Japan Inc."                 , 97, 24, 30,  ""},
	{"MPO, France"                        , 97, 25,  0,  "TDK"},
	{"Hitachi Maxell Ltd."                , 97, 25, 20,  ""},
	{"Infodisc Technology Co,Ltd."        , 97, 25, 30,  "MEMOREX, SPEEDA, Lead data"},
	{"Xcitec"                             , 97, 25, 60,  ""},
	{"Fornet International Pte Ltd"       , 97, 26,  0,  "COMPUSA, Cdhouse"},
	{"Postech Corporation"                , 97, 26, 10,  "Mr.Platinum"},
	{"SKC Co Ltd."                        , 97, 26, 20,  "Infinite"},
	{"Fuji Photo Film Co,Ltd."            , 97, 26, 40,  ""},
	{"Lead Data Inc."                     , 97, 26, 50,  "SONY, Gigastorage, MIRAGE"},
	{"CMC Magnetics Corporation"          , 97, 26, 60,  "Daxon, Verbatim, Memorex, Bi-Winner, PLEXTOR, YAMAHA, Melody, Office DEPOT, Philips, eMARK, imation, HyperMedia, Samsung, Shintaro, Techworks"},
	{"Ricoh Company Limited"              , 97, 27,  0,  "Sony, Digital Storage, Csita"},
	{"Plasmon Data Systems Ltd"           , 97, 27, 10,  "Ritek, TDK, EMTEC, ALPHAPET, MANIA"},
	{"Princo Corporation"                 , 97, 27, 20,  ""},
	{"Pioneer"                            , 97, 27, 30,  ""},
	{"Eastman Kodak Company"              , 97, 27, 40,  ""},
	{"Mitsui Chemicals Inc."              , 97, 27, 50,  "MAM-A, TDK"},
	{"Ricoh Company Limited"              , 97, 27, 60,  "Ritek"},
	{"Gigastorage Corporation"            , 97, 28, 10,  "MaxMax, Nan-Ya"},
	{"Multi Media Masters&Machinary SA"   , 97, 28, 20,  "King, Mmirex"},
	{"Ritek Corp"                         , 97, 31,  0,  "TDK"},
	{"Grand Advance Technology Sdn. Bhd." , 97, 31, 30,  ""},
	{"TDK Corporation"                    , 97, 32, 00,  ""},
	{"Prodisc Technology Inc."            , 97, 32, 10,  "Smartbuy, Mitsubishi, Digmaster, LG, Media Market"},
	{"Mitsubishi Chemical Corporation"    , 97, 34, 20,  "YAMAHA, Verbatim"},
	{"Mitsui Chemicals Inc."              , 97, 48, 50, ""},
	{"TDK Corporation"                    , 97, 49,  0, ""},
	{"", 0, 0, 0, ""}
	};

	int i, f_li_0;
	char buf[1024];
	char *result = NULL;

	if (m_li == 0 && s_li == 2 && f_li == 0) {
		result = strdup("(no manufacturer code)");
		return result;
	}
	f_li_0 = f_li - (f_li % 10);
	for (i = 0; mid_list[i].manufacturer[0]; i++) {
		if (m_li == mid_list[i].m_li &&
		    s_li == mid_list[i].s_li &&
		    (f_li_0 == mid_list[i].f_li || f_li == mid_list[i].f_li))
	break;
	}
	if (mid_list[i].manufacturer[0] == 0) {
		sprintf(buf, "Unknown CD manufacturer. Please report code '%2.2dm%2.2ds%2.2df/%2.2dm%2.2ds%2.2df', the human readable brand, size, and speed to scdbackup@gmx.net.", m_li, s_li, f_li, m_lo, s_lo, f_lo);
		result = strdup(buf);
		return result;
	}

	/* Compose, allocate and copy result */
	if ((flag & 1) && mid_list[i].other_brands[0]) {
		sprintf(buf, "%s  (aka %s)",
			mid_list[i].manufacturer, mid_list[i].other_brands);
		result = strdup(buf);
	} else
		result = strdup(mid_list[i].manufacturer);
	return result;
}


/* ts A90904 */
struct dvd_mid_record {
	char *mc1;
	int mc1_sig_len;
	char *manufacturer;
};
typedef struct dvd_mid_record dvd_mid_record_t;

/* ts A90904 */
char *burn_guess_manufacturer(int prf,
			 char *media_code1, char *media_code2, int flag)
{
	int i, l = 0, m_li, s_li, f_li, m_lo, s_lo, f_lo;
	char buf[1024];
	char *result = NULL, *cpt;

	/* Important Note: media_code1 and media_code2 are supposed to be
	                   encoded by burn_util_make_printable_word().
	                   Especially: ' ' -> '_' , {"_%/" unprintables -> %XY)
	*/
	static dvd_mid_record_t mid_list[]= {
	{"AML",         8, "UML"},
	{"BeAll",       5, "BeAll Developers, Inc."},
	{"CMC",         3, "CMC Magnetics Corporation"},
	{"DAXON",       5, "Daxon Technology Inc. / Acer"},
	{"Daxon",       5, "Daxon Technology Inc. / Acer"},
	{"FUJI",        4, "Fujifilm Holdings Corporation"},
	{"INFODISC",    8, "New Star Digital Co., Ltd."},
	{"INFOME",      6, "InfoMedia Inc."},
	{"ISMMBD",      6, "Info Source Multi Media Ltd."},
	{"JVC",         3, "JVC Limited"},
	{"KIC01RG",     7, "AMC"},
	{"LD",          8, "Lead Data Inc."},
	{"LGE",         3, "LG Electronics"},
	{"MAM",         8, "Mitsui Advanced Media, Inc. Europe"},
	{"MAXELL",      6, "Hitachi Maxell Ltd."},
	{"MBI",         3, "Moser Baer India Limited"},
	{"MCC",         8, "Mitsubishi Chemical Corporation"},
	{"MCI",         8, "Mitsui Chemicals Inc."},
	{"MEI",         3, "Panasonic Corporation"},
	{"MILLEN",      8, "Millenniata Inc."},
	{"MKM",         3, "Mitsubishi Kagaku Media Co."},
	{"MMC",         8, "Mitsubishi Kagaku Media Co."},
	{"MXL",         8, "Hitachi Maxell Ltd."},
	{"NANYA",       5, "Nan-Ya Plastics Corporation"},
	{"NSD",         8, "NESA International Inc."},
	{"OPTODISC",    8, "Optodisc Technology Corporation"},
	{"OTCBDR",      8, "Optodisc Technology Corporation"},
	{"PHILIP",      8, "Moser Baer India Limited"},
	{"PHILIPS",     8, "Philips"},
	{"PRINCO",      6, "Princo Corporation"},
	{"PRODISC",     7, "Prodisc Technology Inc."},
	{"Prodisc",     7, "Prodisc Technology Inc."},
	{"PVC",         3, "Pioneer"},
	{"RICOHJPN",    8, "Ricoh Company Limited"},
	{"RITEK",       5, "Ritek Corp"},
	{"SONY",        4, "Sony Corporation"},
	{"TDK",         3, "TDK Corporation"},
	{"TTG",         3, "TDK Corporation"},
	{"TTH",         3, "TDK Corporation"},
	{"TY",          8, "Taiyo Yuden Company Limited"},
	{"TYG",         3, "Taiyo Yuden Company Limited"},
	{"UME",         3, "UmeDisc Limited"},
	{"UTJR001",     7, "Unifino Inc."},
	{"VERBAT",      5, "Mitsubishi Kagaku Media Co."},
	{"YUDEN",       5, "Taiyo Yuden Company Limited"},
	{"", 0, ""}
	};

	if (media_code2 != NULL &&
	    (prf == -1 || prf == 0x09 || prf == 0x0A)) {
		if (strlen(media_code2) == 9 && media_code1[0] == '9' &&
			media_code1[2] == 'm' && media_code1[5] == 's' &&
			media_code1[8] == 'f' &&
			strchr(media_code1, '%') == NULL) {
			sscanf(media_code1, "%dm%ds%df", &m_li, &s_li, &f_li);
			sscanf(media_code2, "%dm%ds%df", &m_lo, &s_lo, &f_lo);
			if (m_li >= 96 && m_li <= 97 && m_lo > 0) {
				result = burn_guess_cd_manufacturer(
					m_li, s_li, f_li, m_lo, s_lo, f_lo, 0);
				return result;
			}
		}
	}

	/* DVD-R do not keep manufacturer id apart from media id.
	   Some manufacturers use a blank as separator which would now be '_'.
	*/
	cpt = strchr(media_code1, '_');
	if (cpt != NULL && (prf == -1 || prf ==  0x11 || prf == 0x13 ||
						 prf == 0x14 || prf == 0x15))
		l = cpt - media_code1;

	for (i = 0; mid_list[i].mc1[0]; i++) {
		if (strncmp(mid_list[i].mc1, media_code1,
			 mid_list[i].mc1_sig_len) == 0)
	break;
		if (l > 0)
			if (strncmp(mid_list[i].mc1, media_code1, l) == 0)
	break;
	}
	if (mid_list[i].mc1[0] == 0) {
		sprintf(buf, "Unknown DVD/BD manufacturer. Please report code '%s/%s', the human readable brand, size, and speed to scdbackup@gmx.net.", 
			media_code1, media_code2);
		result = strdup(buf);
		return result;
	}
	result = strdup(mid_list[i].manufacturer);
	return result;
}


/* ts A90905 */
/* Make *text a single printable word */
/* IMPORTANT: text must be freeable memory !
   @param flag bit0=escape '/' too
               bit1=(overrides bit0) do not escape " _/"
*/
int burn_util_make_printable_word(char **text, int flag)
{
	int i, esc_add = 0, ret;
	char *wpt, *rpt, *new_text = NULL;

	if (flag & 2)
		flag &= ~1;

	for (i = 0; (*text)[i]; i++) {
		rpt = (*text) + i;
		if (*rpt < 32 || *rpt > 126 || *rpt == 96 ||
		    ((*rpt == '_' || *rpt == '%') && (!(flag & 2))) ||
		    (*rpt == '/' && (flag & 1)))
			esc_add += 2;
	}
	if (esc_add) {
		new_text = calloc(strlen(*text) + esc_add + 1, 1);
		if (new_text == NULL) {
			ret = -1;
			goto ex;
		}
		wpt = new_text;
		for (i = 0; (*text)[i]; i++) {
			rpt = (*text) + i;
			if (*rpt < 32 || *rpt > 126 || *rpt == 96 ||
			    ((*rpt == '_' || *rpt == '%') && (!(flag & 2))) ||
			    (*rpt == '/' && (flag & 1))) {
				sprintf(wpt, "%%%2.2X", 
				    (unsigned int) *((unsigned char *) rpt));
				wpt+= 3;
			} else
				*(wpt++) = *rpt;
		}
		*wpt = 0;
		free(*text);
		*text = new_text;
	}
	if (!(flag & 2))
		for (i = 0; (*text)[i]; i++)
			if ((*text)[i] == ' ')
				(*text)[i] = '_';
	ret = 1;
ex:
	return ret;
}


/* ts B11216 */
/** Read a line from fp and strip LF or CRLF */
char *burn_sfile_fgets(char *line, int maxl, FILE *fp)
{
	int l;
	char *ret;

	ret = fgets(line, maxl, fp);
	if (ret == NULL)
		return NULL;
	l = strlen(line);
	if (l > 0)
		if (line[l - 1] == '\r')
			line[--l] = 0;
	if (l > 0)
		if (line[l - 1] == '\n')
			line[--l] = 0;
	if(l > 0)
		if(line[l - 1] == '\r')
			line[--l] = 0;
	return ret;
}


char *burn_printify(char *msg)
{
	char *cpt;

	for (cpt = msg; *cpt != 0; cpt++)
		if (*cpt < 32 || *cpt > 126)
			*cpt = '#';
	return msg;
}


/* ts B30521 */
void burn_int_to_lsb(int val, char *target)
{
	unsigned char *buf;

	buf = (unsigned char *) target;
	buf[0] = val & 0xff;
	buf[1] = (val >> 8) & 0xff;
	buf[2] = (val >> 16) & 0xff;
	buf[3] = (val >> 24) & 0xff;
}


/* ts B30609 */
double burn_get_time(int flag)
{
	int ret;
	struct timeval tv;

#ifdef Libburn_use_clock_gettime_monotoniC
#ifdef _POSIX_TIMERS
#ifdef _POSIX_MONOTONIC_CLOCK

	/* Enable by
	     export CFLAGS=-DLibburn_use_clock_gettime_monotoniC
	     export LIBS=-lrt
	     ./configure ... && make clean && make
	*/

	struct timespec tp;
	ret = clock_gettime(CLOCK_MONOTONIC, &tp);
	if (ret == 0)
		return ((double) tp.tv_sec) + ((double) tp.tv_nsec) * 1.0e-9;

#endif /* _POSIX_MONOTONIC_CLOCK */
#endif /* _POSIX_TIMERS */
#endif /* Xorriso_use_clock_gettime_monotoniC */

	ret = gettimeofday(&tv, NULL);
	if (ret == 0)
		return ((double) tv.tv_sec) + ((double) tv.tv_usec) * 1.0e-6;
	return (double) time(NULL);
}

/* ts B40609 */
off_t burn_sparse_file_addsize(off_t write_start, struct stat *stbuf)
{
	off_t add_size;

	add_size = stbuf->st_blocks * (off_t) 512;
	if (add_size < stbuf->st_size) {
		/* Sparse file */
		if (write_start < stbuf->st_size) {
			/* Might write into sparse gaps */
			if (write_start > add_size)
				add_size = write_start;
		} else {
			/* Will not write into sparse area */
			add_size = stbuf->st_size;
		}
	}
	return add_size;
}

